### ISSUE TITLE
GPflow 2.0 & tf.saved_model

### ISSUE THREAD

**Issue by marcoadurno** [GPflow/GPflow#1127]
System information:

* Python version: 3.7
* TensorFlow installed from (source or binary): binary
* TensorFlow version (use command below): 'v2.0.0-rc2-26-g64c3d382ca', '2.0.0'
* GPflow installed from (source or binary): source
* GPflow version: develop_2.0, f070a8b

I tried to export a GPflow model using the `tf.saved_model` API by simply decorating the `predict_f` method with the `@tf.function` decorator. The export fails.

I'd expect the export to succeed.

The first obstacle seems to be the Parameter class. Even the following simple export fails:

```
import gpflow
import tensorflow as tf

x = tf.Variable(1.)
p = gpflow.base.Parameter(x)
tf.saved_model.save(p, "test")
```
Simply exporting the `tf.Variable` does work, exporting the `gpflow.base.Parameter` throws (Traceback below).

I've read the [tf.Module rfc](https://github.com/tensorflow/community/blob/master/rfcs/20190117-tf-module.md) and I gather that this is also used by GPflow [gpflow.base.Module](https://github.com/GPflow/GPflow/blob/develop-2.0/gpflow/base.py#L29) and [gpflow.base.Parameter](https://github.com/GPflow/GPflow/blob/develop-2.0/gpflow/base.py#L39). 

Is there any clear reasons why the export fails?

<pre>
Traceback (most recent call last):
  File "test.py", line 10, in <module>
    tf.saved_model.save(p, "test")
  File "C:\Users\madurno\AppData\Local\Continuum\anaconda3\envs\gpflow\lib\site-packages\tensorflow_core\python\saved_model\save.py", line 893, in save
    meta_graph_def, saveable_view, signatures, options.namespace_whitelist)
  File "C:\Users\madurno\AppData\Local\Continuum\anaconda3\envs\gpflow\lib\site-packages\tensorflow_core\python\saved_model\save.py", line 559, in _fill_meta_graph_def
    object_map, resource_map, asset_info = saveable_view.map_resources()
  File "C:\Users\madurno\AppData\Local\Continuum\anaconda3\envs\gpflow\lib\site-packages\tensorflow_core\python\saved_model\save.py", line 269, in map_resources
    new_variable = resource_variable_ops.copy_to_graph_uninitialized(obj)
  File "C:\Users\madurno\AppData\Local\Continuum\anaconda3\envs\gpflow\lib\site-packages\tensorflow_core\python\ops\resource_variable_ops.py", line 1949, in copy_to_graph_uninitialized
    constraint=var._constraint,
AttributeError: 'Parameter' object has no attribute '_constraint'
</pre>

**Comment by marcoadurno** [comment]
Hello,

A bit of an update. The reason why it throws is because the Parameter class doesn't mirror all the properties of the Variable class. From what I can see the missing ones are: ` _constraint`, `shared_name`, `synchronization`, `aggregation` so the following patch is enough to make the code not throw:
```
--- a/gpflow/base.py
+++ b/gpflow/base.py
@@ -101,6 +101,22 @@ class Parameter(tf.Module):
     def trainable(self):
         return self._unconstrained.trainable

+    @property
+    def _constraint(self):
+        return self._unconstrained._constraint
+
+    @property
+    def _shared_name(self):
+        return self._unconstrained._shared_name
+
+    @property
+    def synchronization(self):
+        return self._unconstrained.synchronization
+
+    @property
+    def aggregation(self):
+        return self._unconstrained.aggregation
+
```
Saving the parameter no longer throws, however it is not possible to load the Parameter back.

### Saving a tf.Variable:
```
    x = tf.Variable([1.], dtype=tf.float64, name='variable')
    m = tf.Module()
    m.x = x
    tf.saved_model.save(m, "test_variable")
    m = tf.saved_model.load('./test_variable')
    print(m.x.read_value())
```
yields `tf.Tensor([1.], shape=(1,), dtype=float64)`

### Saving a gpflow.base.Parameter:
```
    x = tf.Variable([1.], dtype=tf.float64, name='variable')
    p = gpflow.base.Parameter(x, name='parameter')
    tf.saved_model.save(p, "test")
    p = tf.saved_model.load('./test')
    print(p._unconstrained.read_value())
```
successfully loads back the model but throws when reading the parameter value:
```
Traceback (most recent call last):
  File ".\test.py", line 19, in <module>
    print(p._unconstrained.read_value())
  File "C:\Users\madurno\AppData\Local\Continuum\anaconda3\envs\gpflow\lib\site-packages\tensorflow_core\python\ops\resource_variable_ops.py", line 630, in read_value
    value = self._read_variable_op()
  File "C:\Users\madurno\AppData\Local\Continuum\anaconda3\envs\gpflow\lib\site-packages\tensorflow_core\python\ops\resource_variable_ops.py", line 608, in _read_variable_op
    self._dtype)
  File "C:\Users\madurno\AppData\Local\Continuum\anaconda3\envs\gpflow\lib\site-packages\tensorflow_core\python\ops\gen_resource_variable_ops.py", line 582, in read_variable_op
    _six.raise_from(_core._status_to_exception(e.code, message), None)
  File "<string>", line 3, in raise_from
tensorflow.python.framework.errors_impl.FailedPreconditionError: Error while reading resource variable variable_218 from Container: localhost. This could mean that the variable was uninitialized. Not found: Resource localhost/variable_218/class tensorflow::Var does not exist. [Op:ReadVariableOp]
```

Is there any obvious reason why this happens?

Best Regards,

Marco

**Comment by condnsdmatters** [comment]
Hi Marco - sorry for the late reply here, thank you for opening the issue! We'll get on looking at this as soon as possible! I think the issue must have slipped under the radar...

**Comment by awav** [comment]
Hello @marcoadurno, can you use checkpointing instead? https://www.tensorflow.org/api_docs/python/tf/train/Checkpoint

**Comment by marcoadurno** [comment]
Hello @awav, yes checkpoint works fine but I need to be able to export the model as saved_model.

I have found a way around the tf bug by replacing the model parameters with variables holding the parameter value at the end of training. You lose the prior and the transform, but at least you can make predictions correctly reusing gpflow code and you can export the model as saved_model.

Having said that, I'm not suggesting we do something along those lines in gpflow as it's a bodge to workaround a TF bug.

Thanks,

Marco

**Comment by mffrank** [comment]
Hi everyone, I am wondering if there are any news on this. I would also need to export models as saved models in my workflow. As I understand checkpointing only allows me to reload models in the same session.

@marcoadurno Could you let me know what exactly you do for your workaround? Are you manually replacing all parameters with variables, or is there a nicer solution?

Thanks in advance,
Max

**Comment by marcoadurno** [comment]
Hi @mffrank ,

Yes, I simply replace all parameters with variables holding the parameter value once training is completed. You can do that in a programmatic way leveraging `gpflow.utilities.parameter_dict`.

Best Regards,

Marco

**Comment by st--** [comment]
@mffrank this is now merged as `gpflow.utilities.freeze()` into develop (PR #1293).

**Comment by mffrank** [comment]
Thanks!

**Comment by st--** [comment]
@marcoadurno @mffrank @cristealab @jajcayn with the changes in #1518 - released as part of GPflow 2.1 - `gpflow.Parameter` is no longer pretending to be a tf.Variable, and `tf.saved_model.save()` works just fine. I've updated the intro_to_gpflow2 notebook in #1586, which will close this issue. Thanks for your patience!

**Comment by st--** [comment]
:tada:
